--- automake-1.16.1/bin/aclocal.in.orig	2018-02-26 20:38:27.000000000 +0000
+++ automake-1.16.1/bin/aclocal.in	2019-08-31 00:21:47.517325614 +0000
@@ -1,10 +1,12 @@
-#!@PERL@ -w
+#!@PERL@
 # -*- perl -*-
 # @configure_input@
 
 eval 'case $# in 0) exec @PERL@ -S "$0";; *) exec @PERL@ -S "$0" "$@";; esac'
     if 0;
 
+$^W = 1;
+
 # aclocal - create aclocal.m4 by scanning configure.ac
 
 # Copyright (C) 1996-2018 Free Software Foundation, Inc.
@@ -27,7 +29,7 @@
 
 BEGIN
 {
-  unshift (@INC, '@datadir@/@PACKAGE@-@APIVERSION@')
+  unshift (@INC, ($ENV{'STAGING_DIR'} ? $ENV{'STAGING_DIR'} . '/../host/share/@PACKAGE@-@APIVERSION@' : '@datadir@/@PACKAGE@-@APIVERSION@'))
     unless $ENV{AUTOMAKE_UNINSTALLED};
 }
 
@@ -68,8 +70,8 @@
 # ACLOCAL_PATH environment variable, and reset with the '--system-acdir'
 # option.
 my @user_includes = ();
-my @automake_includes = ('@datadir@/aclocal-' . $APIVERSION);
-my @system_includes = ('@datadir@/aclocal');
+my @automake_includes = ($ENV{'STAGING_DIR'} ? $ENV{'STAGING_DIR'} . "/../host/share/aclocal-$APIVERSION" : "@datadir@/aclocal-$APIVERSION");
+my @system_includes = ($ENV{'STAGING_DIR'} ? $ENV{'STAGING_DIR'} . '/../host/share/aclocal' : '@datadir@/aclocal');
 
 # Whether we should copy M4 file in $user_includes[0].
 my $install = 0;
--- automake-1.16.1/bin/automake.in.orig	2018-03-08 20:27:05.000000000 +0000
+++ automake-1.16.1/bin/automake.in	2019-08-31 00:20:14.475394924 +0000
@@ -1,10 +1,12 @@
-#!@PERL@ -w
+#!@PERL@
 # -*- perl -*-
 # @configure_input@
 
 eval 'case $# in 0) exec @PERL@ -S "$0";; *) exec @PERL@ -S "$0" "$@";; esac'
     if 0;
 
+$^W = 1;
+
 # automake - create Makefile.in from Makefile.am
 # Copyright (C) 1994-2018 Free Software Foundation, Inc.
 
@@ -31,7 +33,7 @@
 
 BEGIN
 {
-  unshift (@INC, '@datadir@/@PACKAGE@-@APIVERSION@')
+  unshift (@INC, ($ENV{'STAGING_DIR'} ? $ENV{'STAGING_DIR'} . '/../host/share/@PACKAGE@-@APIVERSION@' : '@datadir@/@PACKAGE@-@APIVERSION@'))
     unless $ENV{AUTOMAKE_UNINSTALLED};
 
   # Override SHELL.  This is required on DJGPP so that system() uses
--- automake-1.16.1/lib/Automake/Config.in.orig	2019-08-30 22:17:29.214396399 +0000
+++ automake-1.16.1/lib/Automake/Config.in	2019-08-31 00:23:11.768207277 +0000
@@ -32,7 +32,7 @@
 our $PACKAGE_BUGREPORT = '@PACKAGE_BUGREPORT@';
 our $VERSION = '@VERSION@';
 our $RELEASE_YEAR = '@RELEASE_YEAR@';
-our $libdir = $ENV{"AUTOMAKE_LIBDIR"} || '@datadir@/@PACKAGE@-@APIVERSION@';
+our $libdir = $ENV{'STAGING_DIR'} ? $ENV{'STAGING_DIR'} . '/../host/share/@PACKAGE@-@APIVERSION@' : ($ENV{"AUTOMAKE_LIBDIR"} || '@datadir@/@PACKAGE@-@APIVERSION@');
 
 our $perl_threads = 0;
 # We need at least this version for CLONE support.
