diff --git automake-1.16.1/Makefile.am automake-1.16.1/Makefile.am.orig
index b5ed5eed74..750bc0ec12 100644
--- a/Makefile.am
+++ b/Makefile.am.orig
@@ -124,7 +124,6 @@ EXTRA_DIST += \
 
 # Most work delegated to sub-dir makefile fragments.
 include $(srcdir)/bin/local.mk
-include $(srcdir)/doc/local.mk
 include $(srcdir)/lib/local.mk
 include $(srcdir)/lib/Automake/local.mk
 include $(srcdir)/lib/am/local.mk
